File: 720_longest_word_in_dictionary.py
from typing import List
from collections import deque


class Solution:
    """
        https://leetcode.com/problems/longest-word-in-dictionary/
    """
    """
        // Time Complexity : O(n * w) 
            'n' is the number of words, 'w' is the avg len of the word 
        // Space Complexity : O(w) 
             'w' is the no of unique characters in all words in dic
    """

    class Node:
        def __init__(self):
            self.children = [None] * 26
            self.word = None

    def __init__(self):
        self.root = self.Node()

    def insert(self, word):
        cur = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if cur.children[idx] is None:
                cur.children[idx] = self.Node()
            cur = cur.children[idx]
        cur.word = word

    def longestWord(self, words: List[str]) -> str:
        # insert the words
        for word in words:
            self.insert(word)

        # level order traversal
        queue = deque()
        queue.append(self.root)
        word = ""
        while queue:
            cur = queue.popleft()
            word = cur.word
            for i in reversed(range(26)):
                if cur.children[i] and cur.children[i].word is not None:
                    queue.append(cur.children[i])
        return word

File: 720_LongestWordInDictionary.cpp
// Time Complexity : O(N)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Solution {
public:
    class TrieNode{
        public:
        vector<TrieNode*> child;
        string word;
        TrieNode(){
            child = vector<TrieNode*> (26, nullptr);
            word = "";
        }
    };
    TrieNode* root;
    
    void insert(string word){
        TrieNode* curr = root;
        for(int i=0; i<word.length(); i++){
            char c = word[i];
            if(curr->child[c - 'a'] == NULL)
                curr->child[c - 'a'] = new TrieNode();
            curr = curr->child[c - 'a'];
        }
        curr->word = word;
    }
    
    string longestWord(vector<string>& words) {
        if(words.empty() || words.size() == 0)
            return "";
        root = new TrieNode();
        for(string word : words)
            insert(word);
        //BFS
        queue<TrieNode*> q;
        q.push(root);
        TrieNode* curr = root;
        while(!q.empty()){
            curr = q.front();
            q.pop();
            for(int i=25; i>=0; i--){//to maintain lexicographical order
                if((curr->child[i] != nullptr) && (curr->child[i]->word != ""))
                    q.push(curr->child[i]);
            }
        }
        return curr->word;
    }
};
